src/helper/bit-tools/rect-tool.js

```diff
@@ -0,0 +1,149 @@
+            // Infer that an undo occurred and get back the active rect
+            this.rect = selectedItems[0];
+        }
+    }
+    setColor (color) {
+        this.color = color;
+    }
+    handleMouseDown (event) {
+        if (event.event.button > 0) return; // only first mouse button
+        this.active = true;
+
+        if (this.boundingBoxTool.onMouseDown(event, false /* clone */, false /* multiselect */, this.getHitOptions())) {
+            this.isBoundingBoxMode = true;
+        } else {
+            this.isBoundingBoxMode = false;
+            clearSelection(this.clearSelectedItems);
+            this.commitRect();
+        }
+    }
+    handleMouseDrag (event) {
+        if (event.event.button > 0 || !this.active) return; // only first mouse button
+
+        if (this.isBoundingBoxMode) {
+            this.boundingBoxTool.onMouseDrag(event);
+            return;
+        }
+
+        const dimensions = event.point.subtract(event.downPoint);
+        const baseRect = new paper.Rectangle(event.downPoint, event.point);
+        if (event.modifiers.shift) {
+            baseRect.height = baseRect.width;
+            dimensions.y = event.downPoint.y > event.point.y ? -Math.abs(baseRect.width) : Math.abs(baseRect.width);
+        }
+        if (this.rect) this.rect.remove();
+        this.rect = new paper.Shape.Rectangle(baseRect);
+        this.rect.fillColor = this.color;
+        
+        if (event.modifiers.alt) {
+            this.rect.position = event.downPoint;
+        } else {
+            this.rect.position = event.downPoint.add(dimensions.multiply(.5));
+        }
+    }
+    handleMouseUp (event) {
+        if (event.event.button > 0 || !this.active) return; // only first mouse button
+        
+        if (this.isBoundingBoxMode) {
+            this.boundingBoxTool.onMouseUp(event);
+            this.isBoundingBoxMode = null;
+            return;
+        }
+
+        if (this.rect) {
+            if (Math.abs(this.rect.size.width * this.rect.size.height) < RectTool.TOLERANCE / paper.view.zoom) {
+                // Tiny shape created unintentionally?
+                this.rect.remove();
+                this.rect = null;
+            } else {
+                // Hit testing does not work correctly unless the width and height are positive
+                this.rect.size = new paper.Point(Math.abs(this.rect.size.width), Math.abs(this.rect.size.height));
+                this.rect.selected = true;
+                this.setSelectedItems();
+            }
+        }
+        this.active = false;
+    }
+    commitRect () {
+        if (!this.rect || !this.rect.parent) return;
+
+        const tmpCanvas = document.createElement('canvas');
+        tmpCanvas.width = getRaster().width;
+        tmpCanvas.height = getRaster().height;
+        const context = tmpCanvas.getContext('2d');
+        context.fillStyle = this.color;
+        drawRect(this.rect, context);
+        getRaster().drawImage(tmpCanvas, new paper.Point());
```

Comment:
Just a question, but why is the temporary canvas needed? Could it be drawn directly on the raster?

Reply:
The raster may already have stuff drawn on it. For instance if you drew a rectangle over the cat, then it might just flood fill part of the cat instead of filling in the rectangle.